src/CodeStyle/Core/CodeFixes/Host/Mef/CodeStyleHostLanguageServices.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Concurrent;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis.Host.Mef;
+
+namespace Microsoft.CodeAnalysis.Host
+{
+    internal sealed partial class CodeStyleHostLanguageServices : HostLanguageServices
+    {
+        private static readonly ConcurrentDictionary<HostLanguageServices, CodeStyleHostLanguageServices> s_mappedLanguageServices =
+            new ConcurrentDictionary<HostLanguageServices, CodeStyleHostLanguageServices>();
+
+        private readonly HostLanguageServices _hostLanguageServices;
+        private readonly HostLanguageServices _codeStyleLanguageServices;
+
+        private CodeStyleHostLanguageServices(HostLanguageServices hostLanguageServices)
+        {
+            _hostLanguageServices = hostLanguageServices;
+
+            var exportProvider = new MefHostExportProvider(hostLanguageServices.Language);
```

Comment:
@sharwell @jasonmalinowski I wasn't able to find a way to avoid MEF here as we need access to language service implementation types/singletons, which are defined in the language specific layers. I am going to retain MEF based discovery for this PR and track removal of MEF in https://github.com/dotnet/roslyn/issues/41462 for a follow-up PR.